core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -748,6 +749,22 @@ class ReplicaManager(val config: KafkaConfig,
 
   def tryCompleteActions(): Unit = defaultActionQueue.tryCompleteActions()
 
+  /**
+   * Append an action to the DelayedActionQueue to complete any pending share fetch requests for the already fetched topic partitions.
+   * @param topicIdPartitions          Topic partitions for which we know that the data has been fetched already
+   * @param groupId                    Share group's id
+   * @param delayedShareFetchPurgatory Purgatory storing the pending delayed share fetch requests
+   */
+  def addCompleteDelayedShareFetchPurgatoryAction(topicIdPartitions: Seq[TopicIdPartition],
+                                                  groupId: String,
+                                                  delayedShareFetchPurgatory: DelayedOperationPurgatory[DelayedShareFetch]): Unit = {
```

Comment:
Just for my knowledge, will it not be better to declare `delayedShareFetchPurgatory` in replica manager itself as like other purgatories and define methods to append the requests? I see a lot purgatories are already defined there.

Reply:
Hi @apoorvmittal10 , we can do it but not sure whether it is good to add more scala code. If we declare it in replicaManager, we'll also have to write the functions to add functionalities for `delayedShareFetchPurgatory.tryCompleteElseWatch` and `delayedShareFetchPurgatory.checkAndComplete` in replicaManager which are going to be utilized in `SharePartitionManager` and `SharePartition`. Additionally, we'll need to pass a replicaManager object to SharePartition class as well. Considering all these factors, I feel it is better if we declare it in within `SharePartitionManager`. Let me know if my thoughts don't make sense.

Reply:
I agree with your thoughts and they does make sense.

But I am skeptical about having a single purgatory out of replica manager is not a good idea, if we do have other instances then it's fine. There are metrics and definite standard code to shutdown purgatory and related instances as boilerplate code in replica manager. I know you could handle them separetely as well but it might be better to have them in single place.

I can see list offset purgatory is added recently as well, so I I think it might be ok to have the scala code in old classes. I am not an expert in the area and leave it to @junrao or @mumrah to decide. I just found the API to pass delayed purgatory in replica manager a bit odd.

Reply:
We should declare the new purgatory inside ReplicaManager along with the existing purgatories. We don't need a reference to this purgatory outside of ReplicaManager, we can just add accessor methods.

See ReplicaManager#maybeAddDelayedProduce and  ReplicaManager#completeDelayedFetchOrProduceRequests for how it could work.

To avoid the dependency added to SharePartition, we can add methods on SharePartitionManager (which has a ReplicaManager ref) to add to or complete the purgatory.

Reply:
To me, it seems it's actually better to leave `delayedShareFetchPurgatory` in `SharePartitionManager`. Unlike the fetch/produce purgatory, most of the interactions on `delayedShareFetchPurgatory` happen inside SharePartitionManager. The only thing relevant to ReplicaManager is that if the HWM advances, we need to trigger a check on delayedShareFetchPurgatory. But that can probably be done by having a customized implementation of `LogOffsetsListener.onHighWatermarkUpdated` that adds a delayed action to the ActionQueue.

Reply:
Thanks @junrao -- I wasn't aware of LogOffsetsListener. That seems like a good approach. I was thinking we would need to integrate in ReplicaManager#completeDelayedFetchOrProduceRequests.